// xmldom-ts/dist/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const attr_1 = require("./attr");
exports.AttrImpl = attr_1.AttrImpl;
const cdata_section_1 = require("./cdata-section");
exports.CDATASectionImpl = cdata_section_1.CDATASectionImpl;
const comment_1 = require("./comment");
exports.CommentImpl = comment_1.CommentImpl;
const document_1 = require("./document");
exports.DocumentImpl = document_1.DocumentImpl;
const document_fragment_1 = require("./document-fragment");
exports.DocumentFragmentImpl = document_fragment_1.DocumentFragmentImpl;
const document_type_1 = require("./document-type");
exports.DocumentTypeImpl = document_type_1.DocumentTypeImpl;
const dom_implementation_1 = require("./dom-implementation");
exports.DOMImplementationImpl = dom_implementation_1.DOMImplementationImpl;
const element_1 = require("./element");
exports.ElementImpl = element_1.ElementImpl;
const mutation_observer_1 = require("./mutation/mutation-observer");
exports.MutationObserverImpl = mutation_observer_1.MutationObserverImpl;
const named_node_map_1 = require("./named-node-map");
exports.NamedNodeMapImpl = named_node_map_1.NamedNodeMapImpl;
const node_1 = require("./node");
exports.NodeImpl = node_1.NodeImpl;
const node_list_1 = require("./node-list");
exports.NodeListImpl = node_list_1.NodeListImpl;
const dom_parser_1 = require("./parser/dom-parser");
exports.DOMParserImpl = dom_parser_1.DOMParserImpl;
const processing_instruction_1 = require("./processing-instruction");
exports.ProcessingInstructionImpl = processing_instruction_1.ProcessingInstructionImpl;
const xml_serializer_1 = require("./serializer/xml-serializer");
exports.XMLSerializerImpl = xml_serializer_1.XMLSerializerImpl;
const text_1 = require("./text");
exports.TextImpl = text_1.TextImpl;
if (globalThis) {
    globalThis.XMLSerializer = xml_serializer_1.XMLSerializerImpl;
    globalThis.DOMImplementation = dom_implementation_1.DOMImplementationImpl;
    globalThis.DOMParser = dom_parser_1.DOMParserImpl;
    globalThis.Attr = attr_1.AttrImpl;
    globalThis.CDATASection = cdata_section_1.CDATASectionImpl;
    globalThis.Comment = comment_1.CommentImpl;
    globalThis.DocumentFragment = document_fragment_1.DocumentFragmentImpl;
    globalThis.DocumentType = document_type_1.DocumentTypeImpl;
    globalThis.Document = document_1.DocumentImpl;
    globalThis.Element = element_1.ElementImpl;
    globalThis.NamedNodeMap = named_node_map_1.NamedNodeMapImpl;
    globalThis.NodeList = node_list_1.NodeListImpl;
    globalThis.Node = node_1.NodeImpl;
    globalThis.ProcessingInstruction = processing_instruction_1.ProcessingInstructionImpl;
    globalThis.Text = text_1.TextImpl;
    globalThis.MutationObserver = mutation_observer_1.MutationObserverImpl;
}
