// xmldom-ts/dist/lib/node-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var NodeTypeTS;
(function (NodeTypeTS) {
    NodeTypeTS[NodeTypeTS["ELEMENT_NODE"] = 1] = "ELEMENT_NODE";
    NodeTypeTS[NodeTypeTS["ATTRIBUTE_NODE"] = 2] = "ATTRIBUTE_NODE";
    NodeTypeTS[NodeTypeTS["TEXT_NODE"] = 3] = "TEXT_NODE";
    NodeTypeTS[NodeTypeTS["CDATA_SECTION_NODE"] = 4] = "CDATA_SECTION_NODE";
    NodeTypeTS[NodeTypeTS["ENTITY_REFERENCE_NODE"] = 5] = "ENTITY_REFERENCE_NODE";
    NodeTypeTS[NodeTypeTS["ENTITY_NODE"] = 6] = "ENTITY_NODE";
    NodeTypeTS[NodeTypeTS["PROCESSING_INSTRUCTION_NODE"] = 7] = "PROCESSING_INSTRUCTION_NODE";
    NodeTypeTS[NodeTypeTS["COMMENT_NODE"] = 8] = "COMMENT_NODE";
    NodeTypeTS[NodeTypeTS["DOCUMENT_NODE"] = 9] = "DOCUMENT_NODE";
    NodeTypeTS[NodeTypeTS["DOCUMENT_TYPE_NODE"] = 10] = "DOCUMENT_TYPE_NODE";
    NodeTypeTS[NodeTypeTS["DOCUMENT_FRAGMENT_NODE"] = 11] = "DOCUMENT_FRAGMENT_NODE";
    NodeTypeTS[NodeTypeTS["NOTATION_NODE"] = 12] = "NOTATION_NODE";
})(NodeTypeTS = exports.NodeTypeTS || (exports.NodeTypeTS = {}));
